scheduler/pkg/store/memory.go

```diff
@@ -310,83 +311,65 @@ func (m *MemoryStore) updateLoadedModelsImpl(
 	if !ok {
 		return nil, fmt.Errorf("failed to find server %s", serverKey)
 	}
+
+	assignedReplicaIds := make(map[int]struct{})
 	for _, replica := range replicas {
-		_, ok := server.replicas[replica.GetReplicaIdx()]
-		if !ok {
+		if _, ok := server.replicas[replica.replicaIdx]; !ok {
 			return nil, fmt.Errorf(
-				"Failed to reserve replica %d as it does not exist on server %s",
+				"failed to reserve replica %d as it does not exist on server %s",
 				replica.GetReplicaIdx(), serverKey,
 			)
 		}
+		assignedReplicaIds[replica.replicaIdx] = struct{}{}
 	}
 
 	if modelVersion.HasServer() && modelVersion.Server() != serverKey {
 		logger.Debugf("Adding new version as server changed to %s from %s", modelVersion.Server(), serverKey)
 		m.addNextModelVersion(model, model.Latest().modelDefn)
-		return m.updateLoadedModelsImpl(modelKey, model.Latest().GetVersion(), serverKey, replicas)
+		modelVersion = model.Latest()
```

Comment:
If you follow the recursive call here, all that changes is the `modelVersion`. This makes the method hard to follow.

Reply:
Oh...  this will also need to be a loop, to ensure the that if a `modelVersion` has a server, it matches the server key... seems unnecessary though.